process/transactionProcessor.go

```diff
@@ -359,16 +359,99 @@ func (tp *TransactionProcessor) TransactionCostRequest(tx *data.Transaction) (*d
+func (tp *TransactionProcessor) GetTransaction(txHash string, withResults bool, withRelayedTxHash string) (*transaction.ApiTransactionResult, error) {
+	txHashToGetFromObservers := txHash
+
+	// if the relayed tx hash was provided, this one should be requested from observers
+	innerTxRequested := len(withRelayedTxHash) == len(txHash)
+	if innerTxRequested {
+		txHashToGetFromObservers = withRelayedTxHash
+	}
+
+	tx, err := tp.getTxFromObservers(txHashToGetFromObservers, requestTypeFullHistoryNodes, withResults)
 	if err != nil {
 		return nil, err
 	}
 
 	tx.HyperblockNonce = tx.NotarizedAtDestinationInMetaNonce
 	tx.HyperblockHash = tx.NotarizedAtDestinationInMetaHash
 
-	return tx, nil
+	if len(tx.InnerTransactions) == 0 {
+		if innerTxRequested {
+			return nil, fmt.Errorf("%w, requested hash %s with relayedTxHash %s, but the relayedTxHash has no inner transaction",
+				ErrInvalidHash, txHash, withRelayedTxHash)
+		}
+
+		return tx, nil
+	}
+
+	convertRelayedTxV3ToNetworkTx(tx)
+
+	// requested a relayed transaction, returning it after scrs were moved
+	if !innerTxRequested {
+		return tx, nil
+	}
+
+	for _, innerTx := range tx.InnerTransactions {
+		if innerTx.Hash == txHash {
+			return innerTx, nil
+		}
+	}
+
+	return nil, fmt.Errorf("%w, but the relayedTx %s has no inner transaction with hash %s",
+		ErrInvalidHash, withRelayedTxHash, txHash)
+}
+
+func convertRelayedTxV3ToNetworkTx(tx *transaction.ApiTransactionResult) {
+	movedSCRs := make(map[string]struct{}, 0)
+	for _, innerTx := range tx.InnerTransactions {
+		for _, scr := range tx.SmartContractResults {
+			if isResultOfInnerTx(tx.SmartContractResults, scr, innerTx.Hash) {
+				innerTx.SmartContractResults = append(innerTx.SmartContractResults, scr)
+				movedSCRs[scr.Hash] = struct{}{}
+			}
+		}
+	}
+
+	if len(movedSCRs) == len(tx.SmartContractResults) {
+		// all scrs were generated by inner txs
+		tx.SmartContractResults = make([]*transaction.ApiSmartContractResult, 0)
+		return
+	}
+
+	numSCRsLeftForRelayer := len(tx.SmartContractResults) - len(movedSCRs)
+	scrsForRelayer := make([]*transaction.ApiSmartContractResult, 0, numSCRsLeftForRelayer)
+	for _, scr := range tx.SmartContractResults {
+		_, wasMoved := movedSCRs[scr.Hash]
+		if !wasMoved {
+			scrsForRelayer = append(scrsForRelayer, scr)
+		}
+	}
+
+	tx.SmartContractResults = scrsForRelayer
+}
+
+func isResultOfInnerTx(allScrs []*transaction.ApiSmartContractResult, currentScr *transaction.ApiSmartContractResult, innerTxHash string) bool {
+	if currentScr.PrevTxHash == innerTxHash {
+		return true
+	}
+
+	parentScr := getParentSCR(allScrs, currentScr.PrevTxHash)
+	if check.IfNilReflect(parentScr) {
+		return false
+	}
+
+	return isResultOfInnerTx(allScrs, parentScr, innerTxHash)
+}
+
+func getParentSCR(allScrs []*transaction.ApiSmartContractResult, hash string) *transaction.ApiSmartContractResult {
+	for _, scr := range allScrs {
+		if scr.Hash == hash {
+			return scr
+		}
+	}
```

Comment:
`currentScr` is used for its `PrevTxHash` only. Thus, an optimization would be to hold a map of [previous hash] to [hash]. This way, the repeated loops can be omitted. However, in practice, the optimization would only be marginal. Thus, all right as it is now, as well :+1: